tools/parallelize/parallelize.cpp

```diff
@@ -0,0 +1,203 @@
+// Copyright (c) Microsoft Corporation.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "stljobs.h"
+#include <algorithm>
+#include <condition_variable>
+#include <filesystem>
+#include <limits.h>
+#include <memory>
+#include <mutex>
+#include <optional>
+#include <stdexcept>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string>
+#include <string_view>
+#include <thread>
+#include <vector>
+
+#include <Windows.h>
+
+class tp_wait {
+public:
+    tp_wait(PTP_WAIT_CALLBACK pfnwa, PVOID pv, PTP_CALLBACK_ENVIRON pcbe)
+        : wait(CreateThreadpoolWait(pfnwa, pv, pcbe)) {
+        if (!wait) {
+            api_failure("CreateThreadpoolWait");
+        }
+    }
+
+    tp_wait(const tp_wait&) = delete;
+    tp_wait& operator=(const tp_wait&) = delete;
+
+    void wait_for(const HANDLE waitOn) noexcept {
+        SetThreadpoolWait(wait, waitOn, nullptr);
+    }
+
+    ~tp_wait() {
+        WaitForThreadpoolWaitCallbacks(wait, TRUE);
+        CloseThreadpoolWait(wait);
+    }
+
+private:
+    PTP_WAIT wait{};
+};
+
+class parallelizer {
+public:
+    parallelizer() : subprocesses(std::make_unique<entry[]>(availableConcurrency)) {
+        for (std::size_t idx = 0; idx < availableConcurrency; ++idx) {
+            subprocesses[idx].parent = this;
+        }
+    }
+
+    void add_command(std::wstring&& toRun) {
+        std::lock_guard lck(mtx);
+        const auto oldCommandsSize = commands.size();
+        commands.emplace_back(std::move(toRun), std::nullopt);
+        auto localConcurrency = runningConcurrency.load(std::memory_order_relaxed);
+        if (availableConcurrency == localConcurrency) {
+            return;
+        }
+
+        runningConcurrency.store(localConcurrency + 1, std::memory_order_relaxed);

```

Comment:
Weaker than SC reasoning seems completely uncalled for in something that is manipulating processes and is extremely heavyweight in comparison.

Reply:
The reasoning isn't "weaker than full SC" since the atomic isn't used for any kind of message passing or similar. It's more "this is a regular int that occasionally is read for that `printf`". Note that every access to this variable is relaxed and concurrent updates are protected by `mtx`. (In fact, I want the relaxed to indicate that safety is being provided by some other mechanism here since elsewhere we expect that to only change under lock)